rse-r-package.Rmd

```diff
@@ -0,0 +1,590 @@
+> it is the place to go to find the packages you need.
+> CRAN's rules are famously strict,
+> which ensures that packages run for everyone,
+> but which also makes package development a little more onerous than it might be.
+
+An R package must contain the following files:
+
+-   The text file `DESCRIPTION` (with no suffix) describes what the package does,
+    wrote it,
+    and what other packages it requires to run.
+    We will edit its contents as we go along.
+
+-   `NAMESPACE`,
+    (whose name also has no suffix)
+    contains the names of everything exported from the package
+    (i.e., everything that is visible to the outside world).
+    As we will see,
+    we should leave its management in the hands of RStudio.
+
+-   Just as `.gitignore` tells Git what files in a project to ignore,
+    `.Rbuildignore` tells RStudio which files *shouldn't* be included in the package.
+
+-   All of the R source for our package must go in a directory called `R`;
+    sub-directories below this are not allowed.
+
+-   As you would expect from its name,
+    the `data` directory contains all the data in our package.
+    In order for it to be loadable as part of the package,
+    the data must be saved in R's custom `.rda` format.
+    We will see how to do this below.
+
+-   Manual pages go in the `man` directory.
+    The bad news is that they have to be in a sort-of-LaTeX format
+    that is only a bit less obscure than the runes inscribed on the ancient dagger
+    your colleague brought back from her latest archeological dig.
+    The good news is,
+    we can embed comments in our source code and use a tool called [roxygen2][roxygen2]
+    to extract them and format them as required.
+
+-   The `tests` directory holds the package's unit tests.
+    It should contain files with names like <code>test_<em>some_feature</em>.R</code>,
+    which should in turn contain functions named <code>test_<em>something_specific</em></code>.
+    We'll have a closer look at these in Chapter \@ref(rse-r-testing).
+
+In order to understand the rest of what follows,
+it's important to understand that R packages are distributed as compiled [byte code](glossary.html#byte-code),
+*not* as [source code](glossary.html#source-code) (which is how Python does it)
+or as [binary code](glossary.html#binary-code) (which is how app stores distribute things).
+When a package is built,
+R loads and checks the code,
+then saves the corresponding low-level instructions.
+Our R files should therefore define functions,
+not run commands immediately:
+if they do the latter,
+those commands will be executed every time the script loads,
+which is probably not what users will want.
+
+As a side effect,
+this means that if a package uses `load(something)`,
+then that `load` command is executed *while the package is being compiled*,
+and *not* while the compiled package is being loaded by a user after distribution.
+Thus,
+this simple and rather pointless "package":
```

Comment:
This here seems to read as if you are referring to stringr as pointless, but I assume you mean something else, right?

Reply:
Yes - clarified (it means "the four lines of code below").